Refactor ZHA binary sensors to read from zigpy cache

TheJulianJES opened the pull request:
## Proposed change
Similar to the changes to `sensor.py` in https://github.com/home-assistant/core/pull/43339, this updates ZHA's binary sensors to also just be a representation of zigpy cached attribute values.

#### More info

For sensors which may need to parse the attribute further, the `parse()` method can be overridden.
The `IASZone` binary sensor uses this, as that only needs to "enable the sensor" if either bit 0 or 1 are set. (So the rest needs to be ignored).

These changes align ZHA's binary sensors with how normal sensors are implemented (and probably more).

It also fixes an issue where the custom binary sensors ("Aqara Pet feeder sensors", and so on) did not set their state on pairing (it was correctly read and initialized, but as no attribute report was sent, `async_set_state()` was never called).

I haven't tested it yet, but this might also fix contact sensors having the wrong state when pairing.

The previous implementation of `async_update()` was also somewhat broken, as it relied on the channel having a `value_attribute` or just defaulted to read the `on_off` attribute otherwise.

#### Saving the `zone_status` command data in cache

Since IASZone attributes aren't reportable per spec, ZHA just receives the "zone status change notification" command and now updates the attribute cache based on that command.

#### Old `value_attribute`

The `value_attribute` value/attribute isn't used in the binary sensor class anymore now, but it still seems to be used in some other places. Although it's likely that this can be removed, as all sensors seem to have `SENSOR_ATTR` set correctly.
(If this can be removed, I wouldn't do it in this PR though)

### Only somewhat related to this PR:

#### PR in quirks repo

A fix for a mixup of `STATE` and `STATUS` is here in the quirks repo: https://github.com/zigpy/zha-device-handlers/pull/2231 (will likely change this a bit after this PR)
Currently, that PR changes it to both update the `ZONE_STATUS` attribute and still send the "zone status change notification" command. As the proper attribute is now updated (and sends an attribute report), the command isn't needed (regardless of the changes in ZHA with binary sensor).
Since ZHA now handles caching the attribute, zha-quirks could just send the command.

#### ZONE_STATE vs ZONE_STATUS
Okay, this is getting somewhat off-topic, but the `IASZoneChannel` has `ZCL_INIT_ATTRS` set like this:
```python
ZCL_INIT_ATTRS = {"zone_status": True, "zone_state": False, "zone_type": True}
```
So `status` (alarm or not) and `type` allow cached reads, but `state` (enrolled or not) doesn't.
Is there also a mix-up here or is this intended?

EDIT: This was a mix-up and a fix for that minor issue is also included in this PR.

## Type of change

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

TheJulianJES committed:
Construct binary sensor state from zigpy cache (WIP)

TheJulianJES committed:
Workaround zha-quirks issue where "MotionWithReset" quirks don't update attribute cache (WIP)

zha-quirks currently has an issue where the ZONE_STATE attribute is updated (when the zone_STATUS changes).
https://github.com/zigpy/zha-device-handlers/pull/2231 is a proper fix for this.

For now, we just update the attribute cache when we get the "zone status update notification" command.

This wasn't noticed before, as the "attribute report signal" was sent from the `cluster_command()` method and the used the provided attribute (in the signal) to update the `_state` value in the binary sensor class.
As we just tell HA to write state again when we get an attribute report now, the ZONE_STATUS attribute is read now (and needs to be correct).

TheJulianJES committed:
Use parse() method of main class for IasZone entity (with stripped bits)

TheJulianJES committed:
Change wording in comment, remove explicitly sending attr signal

(This comment should be removed/changed later anyway)

dmulcahey commented on homeassistant/components/zha/core/channels/security.py:
```diff
@@ -341,11 +341,13 @@ class IASZoneChannel(ZigbeeChannel):
-            self.async_send_signal(
-                f"{self.unique_id}_{SIGNAL_ATTR_UPDATED}", 2, "zone_status", state
-            )
+            self.cluster.update_attribute(2, state)  # update cache (sends signal)
```
Magic #

dmulcahey replied:
I think the update of the attr should be done in quirks and not from HA.

dmulcahey commented:
https://zigbeealliance.org/wp-content/uploads/2021/10/07-5123-08-Zigbee-Cluster-Library.pdf#page524

IASZone attributes are also listed as read and not reportable. I think the commands are the only way to reliably get state changes.

TheJulianJES committed:
Remove note

TheJulianJES committed:
Get zone_status attribute id with zigpy

TheJulianJES committed:
Remove `security.` prefix for `IasZone` import

`AceCluster` was already directly imported and `IasZone` is too now for getting the attribute id

dmulcahey commented on homeassistant/components/zha/core/channels/security.py:
```diff
@@ -343,8 +343,9 @@ def cluster_command(self, tsn, command_id, args):
-                f"{self.unique_id}_{SIGNAL_ATTR_UPDATED}", 2, "zone_status", state
+            # update attribute cache (sends signal)
+            self.cluster.update_attribute(
+                IasZone.attributes_by_name["zone_status"].id, state
```
I don't think the "cooked" value should go in the cache. The value reported by the device should.

dmulcahey replied:
We probably improperly do this in a few spots unfortunately

TheJulianJES replied:
Removed all "bit-stripping" from the `IASZoneChannel`. It's only done in the `IASZone` binary sensor now (`parse()` method).

Like before the changes from this PR, this will likely still call `async_write_ha_state()` for the binary sensor even if one of the relevant bits didn't change.
As this was already the case before (it doesn't really matter that much), if that needs to be cleaned up, I'd try to do that in a separate PR. And with that, maybe clean up the `attribute_updated` methods in all different channels and move the check for specific attributes up to the entity level.

TheJulianJES committed:
Store full zone status attribute in cache

TheJulianJES committed:
Check that non-alarm bits are ignored in IasZone sensor test

TheJulianJES committed:
Re-enable occupancy binary sensor test

This test seems to work fine and I don't see any reason why it was commented out for a while

TheJulianJES commented:
I've also re-enabled the occupancy binary sensor test that was commented out in https://github.com/home-assistant/core/pull/36645.
It seems to work like expected and I guess it was only commented out for testing(?)

Adminiuga commented:
> Okay, this is getting somewhat off-topic, but the IASZoneChannel has ZCL_INIT_ATTRS set like this:

```
ZCL_INIT_ATTRS = {"zone_status": True, "zone_state": False, "zone_type": True}
```
> So status (alarm or not) and type allow cached reads, but state (enrolled or not) doesn't.
Is there also a mix-up here or is this intended? 

Yup, it is definitely a mixup. `zone_status` should be a live read and `zone_state` could be a cached read, as it shouldn't change much. Good catch.

TheJulianJES committed:
Fix cached read mix-up for `zone_status`/`zone_state`

This allows cached reads for `zone_state` (enrolled or not), but forces a new read for `zone_status` (alarm or not).

TheJulianJES commented:
Included a commit to fix the mix-up for cached reads. In my testing, everything still works as expected.
